proposals/0000-splice-imports.rst

```diff
@@ -0,0 +1,1113 @@
+statements may not be used inside of top-level splices, nor within quotes.
+
+For example, the following is accepted under the default ``ImplicitStagePersistence``,
+but will be rejected under ``NoImplicitStagePersistence``::
+
+   import B (foo)  -- foo :: Q Exp
+   data C = MkC
+
+   quoteC = [| MkC |]  -- Error: MkC defined at level 0 but used at level 1
+   spliceC = $( foo )  -- Error: foo imported at level 0 but used at level -1
+
+``ImplicitStagePersistence`` is enabled by default in all existing language editions.
+
+Under ``NoImplicitStagePersistence`` it is an error to use ``DeriveLift`` on a
+type unless all its definition is imported at both level 0 and level 1.
+This is discussed in more detail in a later section.
+
+
+
+ExplicitLevelImports
+####################
+
+The ``ExplicitLevelImports`` extension introduces two new import modifiers to
+the import syntax, ``splice`` and ``quote``, which control the level at which
+identifiers from the module are brought into scope:
+
+* A ``splice`` import of ``A`` will import all bindings of ``A`` to be used *only* at
+  level -1.
+* A ``quote`` import of ``B`` will import all bindings of ``B`` to be used
+  *only* at level 1.
+
+For example, the following is accepted under ``ExplicitLevelImports``::
+
+  import quote Foo (bar) -- bar is introduced at level 1
+  import Foo (baz) -- baz is introduced at level 0
+  import splice Foo (qux) -- qux is introduced at level -1
+
+  foo = baz [| bar |] $(qux)
+
+``ExplicitLevelImports`` implies ``NoImplicitStagePersistence``.  Thus users
+typically need only enable ``ExplicitLevelImports`` (and ``TemplateHaskell``).
+
+
+When a module uses ``TemplateHaskell`` with ``NoImplicitStagePersistence``,
+the module dependencies no longer need
+to be pessimistically compiled and loaded at compile time. Instead, the modules
+that are needed at compile-time versus runtime are determined by the explicit
+``splice`` and ``quote`` imports relative to the module being compiled.
+
+It is permitted to enable both ``ExplicitLevelImports`` and
+``ImplicitStagePersistence`` (provided the latter appears later than the former,
+so it overrides the implied ``NoImplicitStagePersistence``). This allows
+``splice`` and ``quote`` imports to be used, but ``ImplicitStagePersistence``
+still allows cross-stage persistence (and thus the compiler must still be
+pessimistically assume all modules are needed at all stages). This combination
+is supported to allow gradual migration of code bases following the change, and
+for corner cases such as programmatic code generation, where the programmer may wish to use
+the syntax of ``splice`` and ``quote`` imports without obliging the whole module
+to be level-correct.
+
+
+Syntax for imports
+##################
+
+Under ``ExplicitLevelImports``, the syntax for imports becomes::
+
+  importdecl :: { LImportDecl GhcPs }
+     : 'import' maybe_src maybe_safe optsplice optqualified maybe_pkg modid optqualified maybeas maybeimpspec
```

Comment:
I think we should anticipate putting qualifiers here after the module name too. ImportQualifiedPost is a popular extension and we probably want to support this from the get go.

also curious if modifier syntax was considered?

Reply:
It seems to me that it would be a good follow-up proposal to introduce another extension if there is demand for it. It would increase the scope of the proposal, at the moment we are concentrating very firmly on just the critical parts which are needed. 

I'm not familiar with modifier syntax? Is that the name for something else which is part of the language?

Reply:
Modifiers are [proposal #0370](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0370-modifiers.rst), I believe the implementation is currently going on and adjustments are happening at #681 .

Reply:
Given that the rest of the import syntax mostly uses keywords (`qualified`, `safe`, `hiding`, ...), it would seem strange to me to use modifier syntax just for splice/quote imports. (Of course `{-# SOURCE #-}` imports are already a weird special case...)

It seems reasonable to want `ImportQualifiedPost`-style postfix keywords, though.

Reply:
Indeed, please do not repeat the original `qualified` mistake.

Allowing level only after the module name is IMO best, though allowing it in both positions is also fine.

Reply:
@mpickering you're welcome to stay out of the syntactic debate, but it is a necessary part of the process!

I've updated the [relevant section of the proposal](https://github.com/well-typed/ghc-proposals/blob/wip/level-imports-2024/proposals/0000-splice-imports.rst#syntactic-alternatives) with the suggestions of which I'm aware, and mentioned in the Unresolved Questions section that the committee needs to make a final decision about the preferred syntax.